### PR TITLE
boards: shields: Add support for Adafruit AW9523 board

### PR DESCRIPTION
**Pull request by soburi**
Add support for Adafruit AW9523 GPIO Expander/LED Controller

I have checked this PR with nulceo_h503rb using the equivalent board as a circuit (https://www.waveshare.com/wiki/AW9523B_IO_Expansion_Board) to Adafruit AW9523.

### CODE REVIEW

File: boards/shields/adafruit_aw9523/adafruit_aw9523.overlay [hunk 1]
```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright (c) 2024 TOKITA Hiroshi
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * This shield has the STEMMA-QT/Qwiic connector, but devicetree defines it as
+ * an electrically compatible Grove connector for convenience.
```
Comment (kartben) [inline boards/shields/adafruit_aw9523/adafruit_aw9523.overlay:9]: I appreciate you taking the time to document this "convenience shortcut" you took but I think you would actually need to define a proper connector + conventional aliases for Qwiic since otherwise this will be problematic for people trying to use this shield on board that has both types of connectors IMO

File: boards/shields/adafruit_aw9523/doc/adafruit_aw9523.jpg
Comment (kartben) [inline boards/shields/adafruit_aw9523/doc/adafruit_aw9523.jpg:1]: Use webp format please. Bonus points if you can find a version with transparent background (not mandatory, but nice to have)

File: boards/shields/adafruit_aw9523/doc/index.rst [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+.. figure:: adafruit_aw9523.jpg
+   :align: center
+   :alt: Adafruit AW9523
+
+   Adafruit AW9523 (Credit: Adafruit)
+
+.. note::
+   This shield has the STEMMA-QT/Qwiic connector, but devicetree defines it as
+   an electrically compatible Grove connector.
+
```
Comment (kartben) [inline boards/shields/adafruit_aw9523/doc/index.rst:21]: Drop when new proper connector will be adopted

File: boards/shields/adafruit_aw9523/doc/index.rst [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+       pulled up.
+
+.. [2] If you want to control the reset pin from the SoC, connect it to a GPIO on the SoC
+       and define the ``reset-gpios`` property in an additional overlay.
+
+Requirements
+************
+
+This shield can only be used with a board which provides a configuration for
+STEMMA QT/Qwiic or Grove.
```
Comment (kartben) [inline boards/shields/adafruit_aw9523/doc/index.rst:48]:
```suggestion
This shield can only be used with a board which provides a configuration for
STEMMA QT/Qwiic.
```

File: boards/shields/adafruit_aw9523/doc/index.rst [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+defines node aliases for SPI and GPIO interfaces (see :ref:`shields` for more details).
```
Comment (kartben) [inline boards/shields/adafruit_aw9523/doc/index.rst:49]: What is this supposed to mean? Maybe a copy paste error?

**Review by kartben — CHANGES_REQUESTED** [review]
Haven't reviewed the driver yet

File: boards/shields/adafruit_aw9523/doc/adafruit_aw9523.jpg
Comment (soburi) [reply boards/shields/adafruit_aw9523/doc/adafruit_aw9523.jpg:1]: I changed the format to webp. I think the photos on the developer's page have the least copyright issues to quote, but unfortunately there were no photos with a white background here. Therefore, I did not change this.

File: drivers/mfd/Kconfig.aw9523b [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+config MFD_AW9523B
+	bool "Awinic AW9523B GPIO/LED controller device driver"
+	default y
+	depends on DT_HAS_AWINIC_AW9523B_ENABLED
+	select I2C
+	help
+	  Enable the Awinic AW9523B GPIO/LED controller device driver
+
+config MFD_AW9523B_INIT_PRIORITY
+	int "nPM1300 MFD initialization priority"
```
Comment (nordicjm) [inline drivers/mfd/Kconfig.aw9523b:13]: wrong comment

File: boards/shields/adafruit_aw9523/doc/adafruit_aw9523.webp
Comment (nordicjm) [inline boards/shields/adafruit_aw9523/doc/adafruit_aw9523.webp:1]: Could you delete the background from this (doesn't have to be transparent but remove the outer parts that are not useful for this imaging and making the filesize far larger), then run it through https://tinypng.com/

File: dts/bindings/gpio/awinic,aw9523b-gpio.yaml [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    type: phandle-array
+    description: |
+      Set GPIO connected to the controller INTN pin.
+      This is used for notifying interruption.
+
+  reset-gpios:
+    type: phandle-array
+    description: |
+      Set GPIO connected to the controller RSTN pin.
+      This is used to controlling resets via GPIO.
```
Comment (kartben) [inline dts/bindings/gpio/awinic,aw9523b-gpio.yaml:23]:
```suggestion
      This is used for controlling resets via GPIO.
```

File: dts/bindings/gpio/stemma-qt-connector.yaml [hunk 7]
```diff
@@ -0,0 +1,21 @@
+# Copyright (c) 2024 TOKITA Hiroshi
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+    STEMMA QT is a 1.0mm-JST-SH connector that
+    externalized the I2C bus.
```
Comment (kartben) [inline dts/bindings/gpio/stemma-qt-connector.yaml:6]:
```suggestion
    STEMMA QT is a 1.0mm-JST-SH connector for I2C devices.
```

File: dts/bindings/mfd/awinic,aw9523b.yaml [hunk 8]
```diff
@@ -0,0 +1,14 @@
+# Copyright (c) 2024 TOKITA Hiroshi
+# SPDX-License-Identifier: Apache-2.0
+
+description: Awinic AW9523B
```
Comment (kartben) [inline dts/bindings/mfd/awinic,aw9523b.yaml:4]:
```suggestion
description: Awinic AW9523B 16 multi-function LED driver and GPIO controller
```

File: boards/shields/adafruit_aw9523/doc/index.rst [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+| SDA                      | AW9523B I2C SDA                          |
++--------------------------+------------------------------------------+
+| SCL                      | AW9523B I2C SCL                          |
++--------------------------+------------------------------------------+
+| INT (Pad on board)       | AW9523B Interrupt output [1]_            |
++--------------------------+------------------------------------------+
+| RST (Pad on board)       | AW9523B Reset pin [2]_                   |
++--------------------------+------------------------------------------+
+
+.. [1] To receive interruption, connect the INT pin to the SoC's GPIO and set the connected
```
Comment (kartben) [inline boards/shields/adafruit_aw9523/doc/index.rst:33]:
```suggestion
.. [1] To receive interrupts, connect the INT pin to the SoC's GPIO and set the connected
```

**Review by kartben — CHANGES_REQUESTED** [review]
minor comments - thanks!

File: boards/shields/adafruit_aw9523/doc/index.rst (hunk 3 above)
Comment (soburi) [reply boards/shields/adafruit_aw9523/doc/index.rst]: Fixed it.

File: drivers/mfd/Kconfig.aw9523b (hunk 5 above)
Comment (soburi) [reply drivers/mfd/Kconfig.aw9523b]: Thank you for pointing that out. This is embarrassing.
Fixed it.

File: boards/shields/adafruit_aw9523/doc/adafruit_aw9523.webp
Comment (soburi) [reply boards/shields/adafruit_aw9523/doc/adafruit_aw9523.webp:1]: Due to my circumstances, this is difficult for me to deal with. It passed the CI (<100kb), so I would appreciate it if you could accept it as is.

Background:

My country (Japan) has strict regulations about modifying other people's copyrighted work. If I modify the background, it does not fall under the category of quoting and requires permission.
There are differing opinions on whether copyright law should be treated on a territorial basis, which can create a worrying situation from a compliance perspective.

For these reasons, my only options are to get this approved or submit this commit without including the image.
Another idea is to take a new photo (it will not be beautiful!) to use here.

Comment (nordicjm) [reply boards/shields/adafruit_aw9523/doc/adafruit_aw9523.webp:1]: No worries then, approved

File: dts/bindings/gpio/awinic,aw9523b-gpio.yaml (hunk 6 above)
Comment (soburi) [reply dts/bindings/gpio/awinic,aw9523b-gpio.yaml]: I applied your suggestion.

File: dts/bindings/gpio/stemma-qt-connector.yaml (hunk 7 above)
Comment (soburi) [reply dts/bindings/gpio/stemma-qt-connector.yaml]: ditto

File: dts/bindings/mfd/awinic,aw9523b.yaml (hunk 8 above)
Comment (soburi) [reply dts/bindings/mfd/awinic,aw9523b.yaml]: ditto

File: boards/shields/adafruit_aw9523/doc/index.rst (hunk 9 above)
Comment (soburi) [reply boards/shields/adafruit_aw9523/doc/index.rst]: ditto

**Review by nordicjm — DISMISSED** [review]

File: boards/shields/adafruit_aw9523/doc/adafruit_aw9523.webp
Comment (soburi) [reply boards/shields/adafruit_aw9523/doc/adafruit_aw9523.webp:1]: Thanks!

**Comment by soburi** [comment]
@henrikbrixandersen @mnkp 

> re-assigning to GPIO collaborators as it would be great to get this bless from a GPIO driver standpoint, as this is a much more substantial part of the PR than the shield bits :)

It would be helpful if you could also take a look at it from the perspective of whether the test can be easily applied to other GPIO Expanders.

**Comment by kartben** [comment]
This might not get a review from GPIO collaborators after all. Maybe @fabiobaltieri will have some cycles to have a look, if it doesn't get merged before.

File: drivers/gpio/Kconfig.aw9523b [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+	select MFD
+	help
+	  Enable the AW9523B GPIO controller.
+
+config GPIO_AW9523B_INIT_PRIORITY
+	int "Init priority"
+	default 80
+	depends on GPIO_AW9523B
+	help
+	  AW9523B GPIO driver initialization priority.
```
Comment (fabiobaltieri) [inline drivers/gpio/Kconfig.aw9523b:18]: can you try and ditch this and just use `CONFIG_MFD_INIT_PRIORITY`? Both here and on the mfd driver. Then subpriorities will control sequencing correctly and we have two less priority options kicking around.

File: drivers/mfd/Kconfig.aw9523b [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+	select I2C
+	help
+	  Enable the Awinic AW9523B GPIO/LED controller device driver
+
+config MFD_AW9523B_INIT_PRIORITY
+	int "Awinic AW9523B GPIO/LED controller initialization priority"
+	default 60
+	depends on MFD_AW9523B
+	help
+	  Initialization priority for AW9523B.
```
Comment (fabiobaltieri) [inline drivers/mfd/Kconfig.aw9523b:17]: see my comment above, I think you can drop this

File: drivers/gpio/gpio_aw9523b.c [hunk 12]
```diff
@@ -0,0 +1,474 @@
...
+	const uint8_t out_high = (flags & GPIO_OUTPUT_INIT_HIGH) ? mask : 0x00;
+	int err;
+
+	/* Can't do I2C operations from an ISR */
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	/* Either INPUT or OUTPUT must be set */
+	if ((!!(flags & GPIO_INPUT) ^ !!(flags & GPIO_OUTPUT)) == 0) {
```
Comment (fabiobaltieri) [inline drivers/gpio/gpio_aw9523b.c:70]: that's a funny way of doing it, how about

`if ((flags & GPIO_INPUT) != 0 && (flags & GPIO_OUTPUT) != 0)`? bit easier to follow

File: drivers/gpio/gpio_aw9523b.c [hunk 13]
```diff
@@ -0,0 +1,474 @@
...
+			uint8_t buf[2];
+
+			/* Read initial pin state */
+			err = i2c_burst_read_dt(&config->i2c, AW9523B_REG_INPUT0, buf, 2);
+			if (err) {
+				LOG_ERR("%s: Read initial pin state failed (%d)", dev->name, err);
+				goto on_error;
+			}
+
+			WRITE_BIT(data->prev_value, pin, !!((buf[1] << 8 | buf[0]) & BIT(pin)));
```
Comment (fabiobaltieri) [inline drivers/gpio/gpio_aw9523b.c:112]: you can use `sys_get_le16` here

File: drivers/gpio/gpio_aw9523b.c [hunk 14]
```diff
@@ -0,0 +1,474 @@
...
+		} else {
+			struct gpio_aw9523b_data *const data = dev->data;
+
+			WRITE_BIT(data->falling_event_pins, pin, 0);
+			WRITE_BIT(data->rising_event_pins, pin, 0);
+		}
+	}
+#endif
+
+	err = i2c_reg_update_byte_dt(&config->i2c, AW9523B_REG_OUTPUT0 + port, mask, out_high);
```
Comment (fabiobaltieri) [inline drivers/gpio/gpio_aw9523b.c:122]: I'd do a define for `AW9523B_REG_OUTPUT0 + port` so you can do `AW9523B_REG_OUTPUT(port)`

File: drivers/gpio/gpio_aw9523b.c [hunk 15]
```diff
@@ -0,0 +1,474 @@
...
+
+	k_sem_take(aw9523b_get_lock(config->mfd_dev), K_FOREVER);
+
+	err = i2c_burst_read_dt(&config->i2c, AW9523B_REG_INPUT0, buf, 1);
+	if (!err) {
+		err = i2c_burst_read_dt(&config->i2c, AW9523B_REG_INPUT1, buf + 1, 1);
+	}
+
+	if (err) {
+		LOG_ERR("%s: Failed to read port status (%d)", dev->name, err);
```
Comment (fabiobaltieri) [inline drivers/gpio/gpio_aw9523b.c:155]: this is quite hard to follow, could it be converted to something simpler? the usual `if (err) { goto unlock_and_return }` type of deal

File: drivers/gpio/gpio_aw9523b.c [hunk 16]
```diff
@@ -0,0 +1,474 @@
...
+	} else if (mode != READ) {
+		gpio_port_value_t old_value = (buf[1] << 8 | buf[0]);
```
Comment (fabiobaltieri) [inline drivers/gpio/gpio_aw9523b.c:157]: use `sys_get_le16`

File: drivers/gpio/gpio_aw9523b.c [hunk 17]
```diff
@@ -0,0 +1,474 @@
...
+
+		if (mode == WRITE) {
+			new_value = (old_value & ~mask) | (*value & mask);
+		} else {
+			new_value = (old_value & ~mask) | (~old_value & mask);
+		}
+
+		if (new_value != old_value) {
+			buf[0] = (new_value >> 0) & 0xFF;
+			buf[1] = (new_value >> 8) & 0xFF;
```
Comment (fabiobaltieri) [inline drivers/gpio/gpio_aw9523b.c:168]: `sys_put_le16` or something  :-)

File: drivers/gpio/gpio_aw9523b.c [hunk 18]
```diff
@@ -0,0 +1,474 @@
...
+			err = i2c_burst_write_dt(&config->i2c, AW9523B_REG_OUTPUT0, buf, 2);
```
Comment (fabiobaltieri) [inline drivers/gpio/gpio_aw9523b.c:169]: `err = i2c_burst_write_dt(&config->i2c, AW9523B_REG_OUTPUT0, buf, sizeof(buf));`

less likely to be copypasted somewhere incorrectly, I usually like all write to refer the source structure if possible, makes it obvious to see that it's right

File: drivers/gpio/gpio_aw9523b.c [hunk 19]
```diff
@@ -0,0 +1,474 @@
...
+			if (err) {
+				LOG_ERR("%s: Failed to set port (%d)", dev->name, err);
+			}
+		}
+	}
+
+	k_sem_give(aw9523b_get_lock(config->mfd_dev));
+
+	if (err == 0 && mode == READ) {
+		*value = (buf[1] << 8 | buf[0]);
```
Comment (fabiobaltieri) [inline drivers/gpio/gpio_aw9523b.c:179]: `sys_get_le16`, but also yeah it's hard to track what code path executes this, if you could refactor it to make all the breanches and exit paths obvious it'd be great

File: drivers/gpio/gpio_aw9523b.c [hunk 20]
```diff
@@ -0,0 +1,474 @@
...
+}
+
+static int gpio_aw9523b_port_set_bits_raw(const struct device *dev, gpio_port_pins_t pins)
+{
+	return gpio_aw9523b_port_read_write_toggle(dev, pins, &pins, WRITE);
+}
+
+static int gpio_aw9523b_port_clear_bits_raw(const struct device *dev, gpio_port_pins_t pins)
+{
+	gpio_port_value_t zero = 0;
```
Comment (fabiobaltieri) [inline drivers/gpio/gpio_aw9523b.c:203]: `const`? (don't think it makes any difference but why  not)

File: drivers/gpio/gpio_aw9523b.c [hunk 21]
```diff
@@ -0,0 +1,474 @@
...
+	uint8_t buf[2];
+	int err;
+
+	/* read INPUT register to deassert INTN */
+	err = i2c_burst_read_dt(&config->i2c, AW9523B_REG_INPUT0, buf, 1);
+	if (err) {
+		LOG_ERR("%s: Failed to read INPUT0 %d", data->dev->name, err);
+	}
+
+	err = i2c_burst_read_dt(&config->i2c, AW9523B_REG_INPUT1, buf + 1, 1);
```
Comment (fabiobaltieri) [inline drivers/gpio/gpio_aw9523b.c:231]: `&buf[1]`? personal preference, up to you, for some reasons I  find it more obvious when dereferencing intermediate buffer points

Also, the registeres are after one another, can't you just do `i2c_burst_read_dt(&config->i2c, AW9523B_REG_INPUT0, buf, 2);`?

File: drivers/gpio/gpio_aw9523b.c [hunk 22]
```diff
@@ -0,0 +1,474 @@
...
+	if (err) {
+		LOG_ERR("%s: Failed to read INPUT1 %d", data->dev->name, err);
+	}
+
+	value = buf[1] << 8 | buf[0];
```
Comment (fabiobaltieri) [inline drivers/gpio/gpio_aw9523b.c:236]: sys_get etc etc, you got the idea

File: drivers/gpio/gpio_aw9523b.c [hunk 23]
```diff
@@ -0,0 +1,474 @@
...
+
+	if (!n_int_en) {
+		/* Read initial pin state */
+		err = i2c_burst_read_dt(&config->i2c, AW9523B_REG_INPUT0, buf, 2);
+		if (err) {
+			LOG_ERR("%s: Failed to read initial pin state (%d)", dev->name, err);
+			return err;
+		}
+
+		WRITE_BIT(data->prev_value, pin, !!((buf[1] << 8 | buf[0]) & BIT(pin)));
```
Comment (fabiobaltieri) [inline drivers/gpio/gpio_aw9523b.c:288]: same as the similar one somewhere above

**Comment by fabiobaltieri** [comment]
@erwango can you take a look at the shield part of this?

File: drivers/gpio/gpio_aw9523b.c (hunk 21 above)
Comment (soburi) [reply drivers/gpio/gpio_aw9523b.c:231]: The operation here is special; if the state of pins 0-7 changes, the interrupt will not be unlatched unless the address of INPUT0 is read, and there is the same relation between 8-15 pins and INPUT1.
At this point, we know that an interrupt has occurred, but we do not know on which side it occurred, so we need to read INPUT0 and INPUT1 independently to clear the interrupt.

Comment (fabiobaltieri) [reply drivers/gpio/gpio_aw9523b.c:231]: got it, thanks for explaining, consider adding a comment for future reference

Comment (soburi) [reply drivers/gpio/gpio_aw9523b.c:231]: OK, I'll address it.

File: drivers/gpio/Kconfig.aw9523b (hunk 10 above)
Comment (soburi) [reply drivers/gpio/Kconfig.aw9523b]: Thank you for your suggestion. This is the ideal solution.

File: drivers/mfd/Kconfig.aw9523b (hunk 11 above)
Comment (soburi) [reply drivers/mfd/Kconfig.aw9523b]: ditto.

File: drivers/gpio/gpio_aw9523b.c (hunk 16 above)
Comment (soburi) [reply drivers/gpio/gpio_aw9523b.c]: I fixed it.

File: drivers/gpio/gpio_aw9523b.c (hunk 17 above)
Comment (soburi) [reply drivers/gpio/gpio_aw9523b.c:191]: I fixed it and other similars.

File: drivers/gpio/gpio_aw9523b.c (hunk 14 above)
Comment (soburi) [reply drivers/gpio/gpio_aw9523b.c]: I fixed it.

File: drivers/gpio/gpio_aw9523b.c (hunk 15 above)
Comment (soburi) [reply drivers/gpio/gpio_aw9523b.c]: I've reorganized this entire function. I think this is getting better.

File: drivers/gpio/gpio_aw9523b.c (hunk 19 above)
Comment (soburi) [reply drivers/gpio/gpio_aw9523b.c]: I've reorganized this entire function. I think this is getting better.

File: drivers/gpio/gpio_aw9523b.c (hunk 18 above)
Comment (soburi) [reply drivers/gpio/gpio_aw9523b.c]: I fixed it and other similars.

File: drivers/gpio/gpio_aw9523b.c (hunk 12 above)
Comment (soburi) [reply drivers/gpio/gpio_aw9523b.c]: Here, I do XOR operations. The equivariant equation is $`{\displaystyle P\veebar Q=(P\lor Q)\land (\lnot P\lor \lnot Q)}`$
I corrected these lines following the equation.

File: dts/bindings/gpio/stemma-qt-connector.yaml [hunk 24]
```diff
@@ -0,0 +1,20 @@
...
+        1  SDA
+        -  VCC (3-5V)
+        -  GND
+
+    STEMMA, Grove, Qwiic, and Gravity connectors are similar to this,
+    but the connector form and VCC voltage are different.
+    Please see
+    https://learn.adafruit.com/introducing-adafruit-stemma-qt/stemma-qt-comparison#quick-comparison-3035247
+
+compatible: "stemma-qt-connector"
```
[the remark this answers is not in the document]
Comment (soburi) [reply dts/bindings/gpio/stemma-qt-connector.yaml:18]: This point will likely cause some debate, but I think keeping the brands and trademarks as they are is best.
This is a kind of guarantee of interoperability, so I think following it would be less confusing.

As I wrote in the commit message, STEMMA QT is only for I2C, and this yaml file is merely a formal definition and doesn't practically have much meaning.
As for STEMMA QT, I think interoperability will be ensured by the naming convention for the labels assigned to i2c.

STEMMA QT and Qwiic have an inclusion relationship in their specifications, and STEMMA QT includes the Qwiic specifications.

In such cases, I think it would be better to create two aliases on the board side, that is,

```
stemma_qt_i2c: &i2c0 { }
qwiic_qt_i2c: &i2c0 { }
```

like this.

In software terms, the signals are in the same order, so it would be easier to unify them, including Grove. Still, I think it would be better to use their official names to identify them and make clear their interoperability.

[the remark this answers is not in the document]
Comment (fabiobaltieri) [reply dts/bindings/gpio/stemma-qt-connector.yaml:18]: does not seem like doubling up is worth it, I mean a shield will be either stemmma/qwiic or grove or the other one but the stemma/qwiic ones interoperate, let's pick one and go with it, really either is good

or rather is either is bad, can't believe they did all this work to do the branding and didn't think about adding an interrupt line to the connector, anyway...

File: dts/bindings/gpio/stemma-qt-connector.yaml [hunk 25]
```diff
@@ -0,0 +1,20 @@
+# Copyright (c) 2024 TOKITA Hiroshi
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+    STEMMA QT is a 1.0mm-JST-SH connector for I2C devices.
```
[the remark this answers is not in the document]
Comment (fabiobaltieri) [reply dts/bindings/gpio/stemma-qt-connector.yaml]: did you actually change anything here? still look the same

File: dts/bindings/gpio/stemma-qt-connector.yaml [hunk 26]
```diff
@@ -0,0 +1,20 @@
...
+    The pin assignments are as follows.
+
+        0  SCL
+        1  SDA
+        -  VCC (3-5V)
+        -  GND
```
Comment (fabiobaltieri) [inline dts/bindings/gpio/stemma-qt-connector.yaml:11]: why did you skip the pin numbers for 2 and 3?

File: drivers/gpio/gpio_aw9523b.c [hunk 27]
```diff
@@ -0,0 +1,504 @@
...
+	} else {
+		WRITE_BIT(data->falling_event_pins, pin, trig & GPIO_INT_LOW_0);
+		WRITE_BIT(data->rising_event_pins, pin, trig & GPIO_INT_HIGH_1);
+	}
+
+	k_sem_take(aw9523b_get_lock(config->mfd_dev), K_FOREVER);
+
+	err = i2c_reg_update_byte_dt(&config->i2c, AW9523B_REG_INT(port), mask, n_int_en);
+	if (err) {
+		LOG_ERR("%s: Failed to configure pin interruption (%d)", dev->name, err);
```
Comment (fabiobaltieri) [inline drivers/gpio/gpio_aw9523b.c:307]: you sure you want to go a head and keep running the rest of the code if this fails? seems a bit pointless, the bus is likely gone for good if we get here, and even if the following call works, what good does it do for us? may as well `return err` and `return 0` at the end

File: drivers/gpio/gpio_aw9523b.c [hunk 28]
```diff
@@ -0,0 +1,504 @@
...
+	static struct gpio_aw9523b_data gpio_aw9523b_data##inst;                                   \
+                                                                                                   \
+	IF_ENABLED(DT_INST_PROP_HAS_IDX(inst, int_gpios, 0), (                                     \
+		static void gpio_aw9523b_int_cb##inst(const struct device *dev,                    \
+							     struct gpio_callback *cb,             \
+							     gpio_port_pins_t pins)                \
+		{                                                                                  \
+			k_work_submit(&gpio_aw9523b_data##inst.intr_worker);                       \
+		}                                                                                  \
+	))                                                                                         \
```
Comment (fabiobaltieri) [inline drivers/gpio/gpio_aw9523b.c:482]: I don't think you need to generate this, can you do

```c
static __maybe_unused void gpio_aw9523b_int_handler(const struct device *gpio_dev,
                                   struct gpio_callback *cb,
                                   uint32_t pins)
{       
        struct gpio_aw9523b_data *data = CONTAINER_OF(
                        cb, struct gpio_aw9523b_data, gpio_callback); 
                
        k_work_submit(&data->intr_worker);
}
```
instead? (copied it from input_paw32xx.c)

**Review by fabiobaltieri — COMMENTED** [review]
good stuff, few more comments

File: dts/bindings/gpio/stemma-qt-connector.yaml (hunk 26 above)
Comment (soburi) [reply dts/bindings/gpio/stemma-qt-connector.yaml:11]: This follows the same notation as the grove.
https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/gpio/grove-header.yaml#L9-L12

Other connectors also use this notation.

File: drivers/gpio/gpio_aw9523b.c (hunk 28 above)
Comment (soburi) [reply drivers/gpio/gpio_aw9523b.c:482]: I worried about whether __maybe_unused can be used. It is standardized in C23, not C99.
If I can use it, it is better to use it.

File: dts/bindings/gpio/stemma-qt-connector.yaml (hunk 24 above)
[the remark this answers is not in the document]
Comment (soburi) [reply dts/bindings/gpio/stemma-qt-connector.yaml:18]: Qwiic is 3.3v, and Stemma is 3V - 5V, so the difficulty is that they are only partially "interoperable". Strictly speaking, qwiic devices can be connected to stemma buses, but Stemma devices cannot be connected to qwiic buses. Also, by definition, a 5V stemma bus can be considered, so there is no interoperability with the Qwiic device in this case. It is a challenging task to express that it will work properly if it can be connected (if it can be compiled in software terms).

Since this PR concerns only the device side, I think the current definition is fine.

> or rather is either is bad, can't believe they did all this work to do the branding and didn't think about adding an interrupt line to the connector, anyway...

This is it!

File: drivers/gpio/gpio_aw9523b.c (hunk 27 above)
Comment (soburi) [reply drivers/gpio/gpio_aw9523b.c:307]: As you pointed out, there was a route just below this that could cause the semaphore release to leak, so we addressed that as well.

File: dts/bindings/gpio/stemma-qt-connector.yaml (hunk 26 above)
Comment (fabiobaltieri) [reply dts/bindings/gpio/stemma-qt-connector.yaml:11]: Oh it's the pin id in the nexus node, of course! Ok nevermind sorry for the noise.

File: drivers/gpio/gpio_aw9523b.c (hunk 28 above)
Comment (fabiobaltieri) [reply drivers/gpio/gpio_aw9523b.c:484]: It's pretty widely used in the code base, I wouldn't worry about it.

File: drivers/gpio/gpio_aw9523b.c (hunk 22 above)
Comment (soburi) [reply drivers/gpio/gpio_aw9523b.c]: This was also addressed.

File: drivers/gpio/gpio_aw9523b.c (hunk 23 above)
Comment (soburi) [reply drivers/gpio/gpio_aw9523b.c]: This was also addressed.

File: dts/bindings/gpio/stemma-qt-connector.yaml [hunk 29]
```diff
@@ -2,7 +2,7 @@
 # SPDX-License-Identifier: Apache-2.0
 
 description: |
-    STEMMA QT is a 1.0mm-JST-SH connector for I2C devices.
+    STEMMA QT is a 4-pin JST-SH connector for I2C devices.
```
Comment (fabiobaltieri) [inline dts/bindings/gpio/stemma-qt-connector.yaml]: almost there, one more commit back :-)

File: drivers/gpio/gpio_aw9523b.c (hunk 28 above)
Comment (soburi) [reply drivers/gpio/gpio_aw9523b.c:484]: I got it. I applied your suggestion.

Comment (fabiobaltieri) [reply drivers/gpio/gpio_aw9523b.c:484]: ok I see you added it on other callbacks but would you not use it to replace `gpio_aw9523b_int_cb##inst` with a common one as well?

Comment (soburi) [reply drivers/gpio/gpio_aw9523b.c]: Ah, I misunderstood what you said.
This is better. I've addressed it.

**Review by fabiobaltieri — DISMISSED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Review by kartben — APPROVED** [review]
